File: challenges/job-queue/javascript/solution/exercise.mjs
// Independent JavaScript practice track; no TypeScript build needed.
export class JobQueue {
    visibilityMs;
    maxAttempts;
    clock;
    ready = [];
    head = 0;
    active = new Map();
    dead = [];
    nextJob = 1;
    nextLease = 1;
    constructor(visibilityMs, maxAttempts, clock) {
        this.visibilityMs = visibilityMs;
        this.maxAttempts = maxAttempts;
        this.clock = clock;
    }
    fail(job) {
        if (job.attempts >= this.maxAttempts)
            this.dead.push({ ...job });
        else
            this.ready.push(job);
    }
    refresh() {
        const now = this.clock();
        const expired = [...this.active.entries()]
            .filter(([, lease]) => lease.deadline <= now)
            .sort((a, b) => a[1].deadline - b[1].deadline || a[1].sequence - b[1].sequence);
        for (const [receipt, lease] of expired) {
            this.active.delete(receipt);
            this.fail(lease.job);
        }
    }
    enqueue(payload) {
        this.refresh();
        const jobId = "job-" + this.nextJob++;
        this.ready.push({ job_id: jobId, payload, attempts: 0 });
        return jobId;
    }
    reserve() {
        this.refresh();
        if (this.head === this.ready.length)
            return null;
        const job = this.ready[this.head++];
        if (this.head * 2 >= this.ready.length) {
            this.ready = this.ready.slice(this.head);
            this.head = 0;
        }
        job.attempts++;
        const sequence = this.nextLease++;
        const receipt = "lease-" + sequence;
        this.active.set(receipt, { job, deadline: this.clock() + this.visibilityMs, sequence });
        return { job_id: job.job_id, payload: job.payload, attempt: job.attempts, receipt };
    }
    ack(receipt) {
        this.refresh();
        return this.active.delete(receipt);
    }
    nack(receipt) {
        this.refresh();
        const lease = this.active.get(receipt);
        if (lease === undefined)
            return false;
        this.active.delete(receipt);
        this.fail(lease.job);
        return true;
    }
    deadLetters() {
        this.refresh();
        return this.dead.map(job => ({ ...job }));
    }
}

File: challenges/job-queue/go/solution/exercise.go
package exercise

import (
	"container/list"
	"fmt"
	"refresher/tooling/go/contracts"
	"sort"
)

type job struct {
	id, payload string
	attempts    int
}
type lease struct {
	job      *job
	deadline int64
	sequence int
}
type expiredLease struct {
	receipt string
	lease   lease
}

type JobQueue struct {
	visibilityMS       int64
	maxAttempts        int
	clock              contracts.Clock
	ready              *list.List
	active             map[string]lease
	dead               []contracts.DeadLetter
	nextJob, nextLease int
}

func NewJobQueue(visibilityMS int64, maxAttempts int, clock contracts.Clock) *JobQueue {
	return &JobQueue{
		visibilityMS: visibilityMS, maxAttempts: maxAttempts, clock: clock,
		ready: list.New(), active: make(map[string]lease),
		dead: []contracts.DeadLetter{}, nextJob: 1, nextLease: 1,
	}
}

func (q *JobQueue) fail(j *job) {
	if j.attempts >= q.maxAttempts {
		q.dead = append(q.dead, contracts.DeadLetter{JobID: j.id, Payload: j.payload, Attempts: j.attempts})
	} else {
		q.ready.PushBack(j)
	}
}

func (q *JobQueue) refresh() {
	now := q.clock()
	expired := make([]expiredLease, 0)
	for receipt, l := range q.active {
		if l.deadline <= now {
			expired = append(expired, expiredLease{receipt, l})
		}
	}
	sort.Slice(expired, func(i, j int) bool {
		a, b := expired[i].lease, expired[j].lease
		if a.deadline == b.deadline {
			return a.sequence < b.sequence
		}
		return a.deadline < b.deadline
	})
	for _, e := range expired {
		delete(q.active, e.receipt)
		q.fail(e.lease.job)
	}
}

func (q *JobQueue) Enqueue(payload string) string {
	q.refresh()
	id := fmt.Sprintf("job-%d", q.nextJob)
	q.nextJob++
	q.ready.PushBack(&job{id: id, payload: payload})
	return id
}

func (q *JobQueue) Reserve() *contracts.Delivery {
	q.refresh()
	element := q.ready.Front()
	if element == nil {
		return nil
	}
	j := q.ready.Remove(element).(*job)
	j.attempts++
	sequence := q.nextLease
	q.nextLease++
	receipt := fmt.Sprintf("lease-%d", sequence)
	q.active[receipt] = lease{j, q.clock() + q.visibilityMS, sequence}
	return &contracts.Delivery{JobID: j.id, Payload: j.payload, Attempt: j.attempts, Receipt: receipt}
}

func (q *JobQueue) Ack(receipt string) bool {
	q.refresh()
	if _, exists := q.active[receipt]; !exists {
		return false
	}
	delete(q.active, receipt)
	return true
}

func (q *JobQueue) Nack(receipt string) bool {
	q.refresh()
	l, exists := q.active[receipt]
	if !exists {
		return false
	}
	delete(q.active, receipt)
	q.fail(l.job)
	return true
}

func (q *JobQueue) DeadLetters() []contracts.DeadLetter {
	q.refresh()
	return append([]contracts.DeadLetter{}, q.dead...)
}
